File: 0917-reverse-only-letters/0917-reverse-only-letters.cpp
class Solution {
public:
    string reverseOnlyLetters(string s) {
        int left = 0, right = s.length() - 1;
        while(left < right){
            while(!isalpha(s[left])){
                left++;
                if(left > s.length() - 1) break;
            }
            while(!isalpha(s[right])){
                right--;
                if(right < 0) break;
            }
            if(left > s.length() - 1 || right < 0 || left > right) break;
            cout << left << "," << right << endl;
            swap(s[left++], s[right--]);
        }
        return s;
    }
};

File: 0917-reverse-only-letters/0917-reverse-only-letters.cs
public class Solution {
    public string ReverseOnlyLetters(string s) {
        //  https://leetcode.com/problems/reverse-only-letters/solutions/4553894/c-easy-and-readable-solution-faster-than-96
        StringBuilder sb = new();
        for(int i = s.Length - 1; i >= 0; i--){
            while(sb.Length < s.Length && !IsLetter(s[sb.Length])){
                sb.Append(s[sb.Length]);
            }
            if(IsLetter(s[i])){
                sb.Append(s[i]);
            }
            while(sb.Length < s.Length && !IsLetter(s[sb.Length])){
                sb.Append(s[sb.Length]);
            }
        }
        return sb.ToString();
    }

    //  https://mathmakeworld.tistory.com/89
    private bool IsLetter(in char ch){
        return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z');
    }
}